Docs: add a development guide (#963)
astarte-platform/astarte: doc/pages/development-guide/001-dev_guide.md

# Development Guide

Welcome to Astarte! To learn more about contributing to the [Astarte project](https://www.github.com/astarte-platform/astarte), check out the [Contributor's Guide](https://github.com/astarte-platform/astarte/blob/master/CONTRIBUTING.md).

> This guide is limited to what needs to be done in order to write and test changes to Astarte. Therefore, it assumes some basic knowledge:

Reviewer: ```suggestion
This guide is focused on writing and testing changes to Astarte. It assumes basic knowledge in the following areas:
```

> - You have read the [contributing guidelines](https://github.com/astarte-platform/astarte/blob/master/CONTRIBUTING.md) to the project

Reviewer: ```suggestion
- You have read the [contributing guidelines](https://github.com/astarte-platform/astarte/blob/master/CONTRIBUTING.md) for the project
```

> - You have some knowledge of the basic Astarte concepts, such as Realms, Interfaces, Devices and Triggers. If you do not, you can start from [here](001-intro_architecture.md)!

Reviewer: ```suggestion
- You are familiar with basic Astarte concepts such as Realms, Interfaces, Devices and Triggers. If you are not, you can learn more on the [documentation](001-intro_architecture.md)!
```

- You have already run at least once [Astarte in 5 minutes](010-astarte_in_5_minutes.md)

> - If you plan to work on the code, you have some knowledge of Elixir and Docker

Reviewer 2: since a setup from zero of elixir and erlang is non-trivial, (docker on the other hand is fairly easy to install), my suggestion is to write down somewhere our current suggested stack (asdf) and the relative plugins (troubleshooting also, for example like the ssl problem, maybe?)
If I remember correctly supported elixir versions also change from one astarte to another, that's another thing that I'd like to know if i started develop with astarte.

Nothing really blocking, as everybody could see the code, but nice to know imho

Reviewer 3: +1 for a section for the environment setup. _maybe nix could be mentioned alongside asdf?_

Author: @eddbbt no need for that SSL troubleshooting section as Astarte 1.0 (i.e. using OTP < 25) is not supported anymore

> - If you plan to work on code, you have some knowledge of Elixir and Docker

Reviewer: ```suggestion
- If you plan to work on the code, you have some knowledge of Elixir and Docker
```

- [Astarte Data Access](https://github.com/astarte-platform/astarte_data_access): contains some interactions with the database that are common to all Astarte services, e.g. queries for interface values.

### Documentation

Documentation is automatically generated from files in the `doc/pages` subdirectory of the [main Astarte repository](https://www.github.com/astarte-platform/astarte) on Github.
It is written in Github-flavored markdown.

### Tooling

The main Astarte repo contains a number of tools that can be used everyday when interacting with Astarte. They are found in the `tools` directory subdirectory of the [main Astarte repository](https://www.github.com/astarte-platform/astarte) on Github:

- Astarte Device Fleet Simulator: create a fleet of virtual devices. Mainly used for load testing
- Astarte E2E: a tool that sends data from a virtual device and checks the resulting value in Astarte using triggers. Mainly used for monitoring
- Astarte Import and Astarte Export: used to load/export data into/from Astarte

## Can I use an Elixir REPL to help me during development?

Sure thing! However, usually Astarte services do need some context around them.

> Refer to the following [”Testing a single component”](#testing-a-single-component) paragraph for dependencies, then you can start your application with an interactive shell by replacing `mix test` with `iex -S mix`

Reviewer 2: Do we have a standard procedure for testing a single component yet? or is it a WIP?

Author: Yup, it's some lines below

Reviewer 2: alright, so anchor ref is broken

Author: No, once it passes through `mix docs` everything will be all right, see e.g. the last line [here](https://docs.astarte-platform.org/astarte/latest/030-interface.html#datastream) which is just
https://github.com/astarte-platform/astarte/blob/fb5657d06ebfc806d944cce354e261c18796960d/doc/pages/architecture/030-interface.md?plain=1#L93

## Ok, I made changes, what now?

Once you got your hands dirty, is time to test them.
You can test changes to a single component, or run E2E tests if more than one service is involved.

Remember that a contribution should always include tests for the new functionality or the fix.
We use the [ExUnit](https://hexdocs.pm/ex_unit/ExUnit.html) framework for unit testing.
Work is under way to add property-based testing too! 

### Testing a single component

In general, you will have to bring up a RabbitMQ and a Scylla instance (not all services need both: API services will not need access to the database).
You can do so by running

```bash
docker run --rm -p 5672:5672 -p 15672:15672 rabbitmq:management
docker run --rm -p 9042:9042 scylladb/scylla
```

Then, you can run test in the component directory with

```bash
RABBITMQ_HOST=localhost CASSANDRA_NODES=localhost mix test --exclude wip
```

> [!NOTE]
> Some services are a bit special (for now!) and might need a little more setup for testing.
> - AppEngine API needs the AMQP exchange `astarte_events` to be declared:
> after having started RabbitMQ, run the following lines to declare it:

> >  ```bash

Reviewer 2: while it's good to give users something to copy and ready to use, we're talking to download and make executable a file from a third party, then delete id. It's probably better to add some context about what is rabbitmqadmin and what could be done with it (link to official docs will suffice)

Reviewer 3: On the other hand, if rabbitmq is started in detached mode and its stdout is assigned to a variable, we can just docker exec into it without needing use wget

```bash
RMQ_CONTAINER=$(docker run -d --rm -p 5672:5672 -p 15672:15672 rabbitmq:management)

docker exec $RMQ_CONTAINER rabbitmqadmin declare exchange name=astarte_events type=direct
```

>   rm rabbitmqadmin
>  ```

> - Pairing needs a CFSSL instance available and exposed on port 8080: you can run
>  ```bash
>  docker run --net=host -p 8080/tcp ispirata/docker-alpine-cfssl-autotest:astarte
>  ```
> 
> and then test using
>  ```bash
>  RABBITMQ_HOST=localhost CASSANDRA_NODES=localhost CFSSL_API_URL=http://localhost:8080 mix test
>  ```

### E2E testing

docker compose is the quick-and-dirty way to spin up a development instance of Astarte.
You can do so just by running `docker compose up -d` in the main Astarte directory.
You can rebuild the service you’re working on with `docker compose build $SERVICE_NAME`,
or edit the `docker-compose.yml` file to change the service image name.

 Some useful commands:

- `docker compose up -d` starts all containers in detached mode
- `docker compose down $SERVICE_NAME -v` stops and cleans the service you’re working on
- `docker compose up $SERVICE_NAME --build` rebuilds the service and restarts it. If you don’t need to tail logs, you may add `-d`

## Useful commands

Now, some commands you might find helpful to set up a realm:

- Create a realm keypair:
`astartectl utils gen-keypair $REALM`
- Create a Realm:

> `astartectl housekeeping realms create $REALM -u http://api.astarte.localhost --housekeeping-key ./compose/astarte-keys/housekeeping_private.pem --realm-private-key $PRIVATE_PEM`

Reviewer 2: user only have ./compose/astarte-keys/housekeeping_private.pem if they have followed astarte in 5 minutes before, better state that?

- Install a list of Interfaces in a Realm:
`astartectl realm-management interfaces sync $(find $INTERFACE_DIRECTORY -name '*.json') -u http://api.astarte.localhost -r $REALM -k $PRIVATE_PEM`
- Create a JWT wth claims on AppEngine, Pairing and RealmManagement APIs:
`astartectl utils gen-jwt all-realm-apis -k $PRIVATE_PEM`
- Start a virtual device:
`docker run --net="host" -e "DEVICE_ID=$(astartectl utils device-id generate-random)" -e "PAIRING_URL=http://api.astarte.localhost/pairing" -e "REALM=$REALM" -e "PAIRING_JWT=$(astartectl utils gen-jwt pairing -k $PRIVATE_PEM)" -e "IGNORE_SSL_ERRORS=true" astarte/astarte-stream-qt5-test:1.0.4`
    - In this case, two interfaces need to be already installed in your realm: `org.astarte-platform.genericsensors.Values` and `org.astarte-platform.genericcommands.ServerCommands`
    - You can find them in the `standard-interfaces` directory in the Astarte repo

> ### Virtual devices

Author: Does it feel better now? I moved some lines around
